### PR TITLE
#20 and some more

### PR DESCRIPTION
**Pull request by dmpe**
Hi @RaymondMcT ,

- [X] In reference to https://github.com/quandl/R-package/issues/20 I have tried to move from RSJSONIO to jsonlite (https://github.com/jeroenooms/jsonlite)
- [ ] Prove that ```jsonlite``` is really faster - in our use case. **Not proved**

**Changes besides that**

- [x] Added 2 tests for search funtionality and "improved" others.
- [x] Add ```GitIgnore``` and ```R-package.Rproj``` files to repository
- [x] Updated ```Changelog```, ```Description```, ```Namespace```, ```Readme``` and ```Man docs```
- [x] New dependency: ```roxygen2``` for easier running of tests and builds from RStudio.
- [x] Some **consistent formatting**, too. BTW: What do you prefer ? I kind of used the default one from RStudio (2 tabs)-everywhere....
   * Add many braces by applying some of the https://google-styleguide.googlecode.com/svn/trunk/Rguide.xml styles
- [x] Passes tests

### CODE REVIEW

**Comment by f789gh** [comment]
Hi, 
Here the @dmpe user. Sorry for confussion with different accounts.  

Per https://twitter.com/hadleywickham/status/551100751217303553 it should be library("package"). I though otherwise. Sorry again, will squash them and improve commits in the comming week.

File: inst/tests/test-get.r [hunk 1]
```diff
@@ -1,46 +1,40 @@
 require("zoo")
 require("xts")
-require("Quandl")
```
Comment (dmpe) [inline inst/tests/test-get.r]: shouldn't that be rather a library() ? Aka: same as in tests/test-all.R

**Comment by dmpe** [comment]
Hi again, 

Regarding performance:
```
library("microbenchmark")

microbenchmark(
  RJSONIO::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/2?&auth_token=GgnxpyUBXHsyQxqp67bY"),
  jsonlite::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/2?&auth_token=GgnxpyUBXHsyQxqp67bY",simplifyDataFrame=FALSE),
  times=1
)
```
I get:
```
Unit: milliseconds
                                                                                                                                 expr      min
                                  RJSONIO::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/2?&auth_token=GgnxpyUBXHsyQxqp67bY") 311.0227
 jsonlite::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/2?&auth_token=GgnxpyUBXHsyQxqp67bY",      simplifyDataFrame = FALSE) 219.4040
       lq     mean   median       uq      max neval
 311.0227 311.0227 311.0227 311.0227 311.0227     1
 219.4040 219.4040 219.4040 219.4040 219.4040     1
```
which (if i am correct) shows that jsonlite() is a little bit faster. However, my previous version had simplifyDataFrame = TRUE so this has been now corrected. I know, it is not huge improvement but still good one.

**Comment by RaymondMcT** [comment]
What version of R are you using?

I just downgraded to 3.1.2 and it benchmarks in favour of RJSONIO
When running 3.2.0 it weighed even more so in favour of RJSONIO

I'm going to continue checking, run it on some other computer since this seems contrary to the blog article and your own analysis.

**Comment by dmpe** [comment]
Hi,
Thanks for looking at that. 

The initial test has been done on latest Ubuntu with 
```
R version 3.1.2 (2014-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Quandl_2.5.1 xts_0.9-7    zoo_1.7-11  

loaded via a namespace (and not attached):
[1] bitops_1.0-6    grid_3.1.2      jsonlite_0.9.14 lattice_0.20-29 RCurl_1.95-4.5  tools_3.1.2   
```
I tested it again: same configuration; "cold system", different datasets:

```
 microbenchmark(
+   RJSONIO::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/3?&auth_token=GgnxpyUBXHsyQxqp67bY"),
+   jsonlite::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/3?&auth_token=GgnxpyUBXHsyQxqp67bY",simplifyDataFrame=FALSE),
+   times=3
+ )
Unit: milliseconds
                                                                                                                                 expr
                                  RJSONIO::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/3?&auth_token=GgnxpyUBXHsyQxqp67bY")
 jsonlite::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/3?&auth_token=GgnxpyUBXHsyQxqp67bY",      simplifyDataFrame = FALSE)
      min       lq     mean   median       uq      max neval
 277.8037 279.0420 279.6060 280.2802 280.5072 280.7341     3
 210.9151 214.8576 217.6153 218.8001 220.9654 223.1306     3
```
I also tested LBMA/GOLD and look 

```
Unit: seconds
                                                                                                                                   expr
                                  RJSONIO::fromJSON("http://www.quandl.com/api/v1/datasets/LBMA/GOLD?&auth_token=GgnxpyUBXHsyQxqp67bY")
 jsonlite::fromJSON("http://www.quandl.com/api/v1/datasets/LBMA/GOLD?&auth_token=GgnxpyUBXHsyQxqp67bY",      simplifyDataFrame = FALSE)
      min       lq     mean   median       uq      max neval
 9.508483 9.614841 9.653441 9.721199 9.725919 9.730640     3
 8.585072 8.761916 8.934331 8.938760 9.108961 9.279162     3
```

Or maybe because I skipped ```nullValue = as.numeric(NA)``` at RJSONIO part ?

```
Unit: seconds
                                                                                                                                   expr
 RJSONIO::fromJSON("http://www.quandl.com/api/v1/datasets/LBMA/GOLD?&auth_token=GgnxpyUBXHsyQxqp67bY",      nullValue = as.numeric(NA))
 jsonlite::fromJSON("http://www.quandl.com/api/v1/datasets/LBMA/GOLD?&auth_token=GgnxpyUBXHsyQxqp67bY",      simplifyDataFrame = FALSE)
       min        lq      mean    median        uq       max neval
 11.094086 11.094086 11.094086 11.094086 11.094086 11.094086     1
  9.083817  9.083817  9.083817  9.083817  9.083817  9.083817     1
```
I dont understand why is there such a big difference. :scream:

Have you run the same test multiple times? 
Nevertheless, if it doesnt prove faster, I will abondon RJSONIO->jsonlite and just commit other changes. 

I will also take a look on it from windows 7 with http://mran.revolutionanalytics.com/download/

**Comment by dmpe** [comment]
Just for information:

```RRO 8.0.1 is based on (and 100% compatible with) version 3.1.2 of the R language, and includes additional capabilities for performance, reproducibility and platform support.```

Platform: latest Win 8.1
```
R version 3.1.2 (2014-10-31)
Platform: x86_64-w64-mingw32/x64 (64-bit)

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Quandl_2.5.1 xts_0.9-7    zoo_1.7-11  

loaded via a namespace (and not attached):
[1] bitops_1.0-6    grid_3.1.2      jsonlite_0.9.14 lattice_0.20-29 RCurl_1.95-4.5  tools_3.1.2    
```
Running same tests gives 
### Test 1 ###
```
microbenchmark(
  RJSONIO::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/3?&auth_token=GgnxpyUBXHsyQxqp67bY"),
  jsonlite::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/3?&auth_token=GgnxpyUBXHsyQxqp67bY",simplifyDataFrame=FALSE),
  times=3
)

Unit: milliseconds
                                                                                                                                 expr
                                  RJSONIO::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/3?&auth_token=GgnxpyUBXHsyQxqp67bY")
 jsonlite::fromJSON("http://www.quandl.com/api/v1/datasets/TESTS/3?&auth_token=GgnxpyUBXHsyQxqp67bY",      simplifyDataFrame = FALSE)
      min       lq     mean   median       uq      max neval
 183.4287 210.5528 249.6499 237.6770 282.7605 327.8440     3
 223.6475 249.9624 262.6534 276.2772 282.1563 288.0354     3
```
Here, ```jsonlite``` is seems to be **slower**. However, on Linux it is (for me) *faster*.


### Test 2 ###
```
Unit: seconds
                                                                                                                                   expr
                                  RJSONIO::fromJSON("http://www.quandl.com/api/v1/datasets/LBMA/GOLD?&auth_token=GgnxpyUBXHsyQxqp67bY")
 jsonlite::fromJSON("http://www.quandl.com/api/v1/datasets/LBMA/GOLD?&auth_token=GgnxpyUBXHsyQxqp67bY",      simplifyDataFrame = FALSE)
       min        lq     mean    median       uq      max neval
 13.402335 13.718847 25.95444 14.035359 32.23049 50.42562     3
  9.420462  9.675147 11.64351  9.929831 12.75503 15.58023     3
```

### Test 3 ###
```
Unit: seconds
                                                                                                                                   expr
 RJSONIO::fromJSON("http://www.quandl.com/api/v1/datasets/LBMA/GOLD?&auth_token=GgnxpyUBXHsyQxqp67bY",      nullValue = as.numeric(NA))
 jsonlite::fromJSON("http://www.quandl.com/api/v1/datasets/LBMA/GOLD?&auth_token=GgnxpyUBXHsyQxqp67bY",      simplifyDataFrame = FALSE)
      min       lq     mean   median       uq      max neval
 37.01320 37.01320 37.01320 37.01320 37.01320 37.01320     1
  9.66007  9.66007  9.66007  9.66007  9.66007  9.66007     1
```


It is really hard for me to compare under such different results. Initially, I though it would be faster (same way as probably you did by reading articles & blog posts), and some early benchmarks supported my idea. 

But now.... it looks much more like *"RJSONIO is faster at reading smaller jsons, whereas jsonlite is better at reading bigger ones. - This is also very much platform specific."*

**Comment by RaymondMcT** [comment]
So I've tested on multiple systems and I cannot reproduce jsonlite being faster

Unix Test
Small
                                                                             expr
                          RJSONIO::fromJSON(response1, simplifyWithNames = FALSE)
 jsonlite::fromJSON(response1, simplifyDataFrame = FALSE, simplifyMatrix = FALSE)
      min       lq     mean   median       uq      max neval
 192.8153 192.8153 192.8153 192.8153 192.8153 192.8153     1
 695.5330 695.5330 695.5330 695.5330 695.5330 695.5330     1
Unit: microseconds

Large
                                                                             expr
                          RJSONIO::fromJSON(response2, simplifyWithNames = FALSE)
 jsonlite::fromJSON(response2, simplifyDataFrame = FALSE, simplifyMatrix = FALSE)
      min       lq     mean   median       uq      max neval
  846.218  846.218  846.218  846.218  846.218  846.218     1
 1372.741 1372.741 1372.741 1372.741 1372.741 1372.741     1

Mac OS X
Small
                                                                             expr
                          RJSONIO::fromJSON(response1, simplifyWithNames = FALSE)
 jsonlite::fromJSON(response1, simplifyDataFrame = FALSE, simplifyMatrix = FALSE)
      min       lq     mean   median       uq      max neval
 100.6617 100.6617 100.6617 100.6617 100.6617 100.6617     1
 379.9979 379.9979 379.9979 379.9979 379.9979 379.9979     1

Large
Unit: microseconds
                                                                             expr
                          RJSONIO::fromJSON(response2, simplifyWithNames = FALSE)
 jsonlite::fromJSON(response2, simplifyDataFrame = FALSE, simplifyMatrix = FALSE)
      min       lq     mean   median       uq      max neval
  842.536  842.536  842.536  842.536  842.536  842.536     1
 1115.730 1115.730 1115.730 1115.730 1115.730 1115.730     1

Could you remove the jsonlite from the pull request? It looks great otherwise and I'd like to merge

**Comment by dmpe** [comment]
Hi Raymond, 

> Could you remove the jsonlite from the pull request? It looks great otherwise and I'd like to merge

Great, thanks. Because I did many squashed commit, I will need some time to get it back. 

However, let me just ask you why do use ```simplifyWithNames = FALSE``` ? Is it a default value ?
Because as of current https://github.com/quandl/R-package/blob/master/R/Quandlapi.R#L65 the package uses ``` nullValue = as.numeric(NA)```; and the documentation doesn't says what is the default value  http://www.inside-r.org/packages/cran/RJSONIO/docs/fromJSON

Amazing

Good anyway- I will get back to RJSONIO.

**Pull request merged**